packages/react/src/module-federation/utils.ts

```diff
@@ -6,17 +6,42 @@ import {
   mapRemotes,
   mapRemotesForSSR,
   ModuleFederationConfig,
-  ProjectConfiguration,
   ProjectGraph,
   readCachedProjectGraph,
   sharePackages,
   shareWorkspaceLibraries,
 } from '@nx/devkit';
+import { readCachedProjectConfiguration } from 'nx/src/project-graph/project-graph';
+
+export function getFunctionDeterminateRemoteUrl(isServer: boolean = false) {
+  const target = isServer ? 'serve-server' : 'serve';
+  const remoteEntry = isServer ? 'server/remoteEntry.js' : 'remoteEntry.js';
+
+  return function (remote: string) {
+    const remoteConfiguration = readCachedProjectConfiguration(remote);
+    const serveTarget = remoteConfiguration?.targets?.[target];
+
+    if (!serveTarget) {
+      throw new Error(
+        `Cannot automatically determine URL of remote (${remote}). Looked for property "host" in the project's "serve-server" target.\n
+      You can also use the tuple syntax in your webpack config to configure your remotes. e.g. \`remotes: [['remote1', 'http://localhost:4201']]\``
+      );
+    }
+
+    const host = serveTarget.options?.host ?? 'http://localhost';
+    const port = serveTarget.options?.port ?? 4201;
```

Comment:
What about using/creating a similar function to the [findNextAvailablePort](https://github.com/nrwl/nx/blob/37038463ed7a33411ade8aaca6510a3c5be56ac3/packages/angular/src/generators/remote/lib/find-next-available-port.ts#L5) to prevent local port collisions?